hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/receiver/kafka/FilteringMessageReceiver.java

```diff
@@ -0,0 +1,65 @@
+package pl.allegro.tech.hermes.consumers.consumer.receiver.kafka;
+
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.consumers.consumer.Message;
+import pl.allegro.tech.hermes.consumers.consumer.filtering.chain.FilterChain;
+import pl.allegro.tech.hermes.consumers.consumer.filtering.chain.FilterChainFactory;
+import pl.allegro.tech.hermes.consumers.consumer.filtering.chain.FilterResult;
+import pl.allegro.tech.hermes.consumers.consumer.filtering.FilteredMessageHandler;
+import pl.allegro.tech.hermes.consumers.consumer.receiver.MessageReceiver;
+
+public class FilteringMessageReceiver implements MessageReceiver {
+    private MessageReceiver receiver;
+    private FilteredMessageHandler filteredMessageHandler;
+    private FilterChainFactory filterChainFactory;
+
+    private volatile boolean dirty = true;
+    private Subscription subscription;
+    private FilterChain filterChain;
+    private boolean consuming = true;
+
+    public FilteringMessageReceiver(MessageReceiver receiver,
+                                    FilteredMessageHandler filteredMessageHandler,
+                                    FilterChainFactory filterChainFactory,
+                                    Subscription subscription) {
+        this.receiver = receiver;
+        this.filteredMessageHandler = filteredMessageHandler;
+        this.filterChainFactory = filterChainFactory;
+        this.subscription = subscription;
+    }
+
+    private boolean filter(final Message message) {
+        FilterResult result = filterChain.apply(message);
+        filteredMessageHandler.handle(result, message, subscription);
+        return result.filtered;
+    }
+
+    @Override
+    public Message next() {
+        Message message;
+        do {
+            updateFilter();
```

Comment:
I will work on this code cause I think it can be more readable but general idea is not to update with every message read (there is a condition hidden in updateFilter) but if we have to update then we're doing that from the same thread that is going to read actual value later.